Class/Practise 2.2/image.cpp

```diff
@@ -0,0 +1,102 @@
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <vector>
+#include <array>
+#include <tuple>
+#include <ranges>
+#include <print>
+
+struct Color {
+    int r, g, b;
+
+    bool operator==(const Color& other) const {
+        return r == other.r && g == other.g && b == other.b;
+    }
+};
+
+using Image = std::array<std::array<Color, 16>, 16>;
+
+bool readImage(const std::string& inputFilename, Image& image) {
+    std::ifstream file(inputFilename);
+    if (!file.is_open()) {
+        std::println("Failed to open file: {}", inputFilename);
+        return false;
+    }
+
+    std::string line;
+    for (int i = 0; i < 16; ++i) {
+        if (!std::getline(file, line)) {
+            std::println("Invalid input format: not enough lines.");
+            return false;
+        }
+
+        std::istringstream lineStream(line);
+        for (int j = 0; j < 16; ++j) {
+            char delimiter;
+            if (!(lineStream >> image[i][j].r >> delimiter >> image[i][j].g >> delimiter >> image[i][j].b) || delimiter != ',') {
+                std::println("Invalid input format at line {}, pixel {}.", i + 1, j + 1);
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+void writeImage(const std::string& outputFilename, const Image& image) {
+    std::ofstream file(outputFilename);
+    for (const auto& row : image) {
+        for (const auto& pixel : row) {
+            file << pixel.r << ',' << pixel.g << ',' << pixel.b << ' ';
+        }
+        file << '\n';
+    }
+}
+
+void applyFavoriteColor(Image& image, const Color& favoriteColor) {
+    for (int i = 0; i < 16; ++i) {
```

Comment:
magic numbers should be replaced with constants or the real image size obtained from the image object.